src/coreclr/vm/comdelegate.cpp

```diff
@@ -189,16 +189,57 @@ class ShuffleIterator
         if (m_currentByteStackIndex < m_argLocDesc->m_byteStackSize)
         {
             const unsigned byteIndex = m_argLocDesc->m_byteStackIndex + m_currentByteStackIndex;
-            index = byteIndex / TARGET_POINTER_SIZE;
+
+#if !defined(TARGET_OSX) || !defined(TARGET_ARM64)
             m_currentByteStackIndex += TARGET_POINTER_SIZE;
 
             // Delegates cannot handle overly large argument stacks due to shuffle entry encoding limitations.
+            index = byteIndex / TARGET_POINTER_SIZE;
             if (index >= ShuffleEntry::REGMASK)
             {
                 COMPlusThrow(kNotSupportedException);
             }
 
-            return -(int)byteIndex;
+            // Only Apple Silicon ABI currently supports unaligned stack argument shuffling
+            _ASSERTE(byteIndex == index * TARGET_POINTER_SIZE);
+            return index;
+#else
+            int bytesRemaining = m_argLocDesc->m_byteStackSize - m_currentByteStackIndex;
```

Comment:
I think this mechanism needs a comment describing what it does. It is not clear to me, as arguments can only be of a  power of two size, but this seems to handle any arbitrary size and splits it to a sequence of TARGET_POINTER_SIZE sized blocks followed by blocks of 4, 2 and 1 bytes sizes (only the ones that are needed to form the target size).

I also wonder if it honors the following from the Apple doc:
> Function arguments may consume slots on the stack that are not multiples of 8 bytes. If the total number of bytes for stack-based arguments is not a multiple of 8 bytes, insert padding on the stack to maintain the 8-byte alignment requirements. 

Imagine a case when all arguments are passed in registers in the source signature, using all the available general purpose registers. The last argument is of a single byte size and we insert a "this" reference as the first argument in the destination signature. The last argument then goes to a stack slot and the stack needs to be aligned to 8.

Reply:
Last I knew the arm64 stack needed to always be 16-byte aligned (to support ld/st pair & simd)

The existing code does not support an argument moving from a register onto the stack.  Only the opposite case is supported (stack -> register). See
https://github.com/dotnet/runtime/blob/c95aa3f48aa591dca870ee00c31c3d8bdc740b5a/src/coreclr/vm/arm64/stubs.cpp#L1720

So the stack only shrinks.... I don't think this code needs to do anything for padding or stack alignment.

I suspect the 3/5/6/7 cases are not really necessary.  It seemed reasonable to just implement them, but I could `_ASSERTE` instead.  My intent is to copy the argument stack size as reported by the `ArgLocDesc`

Reply:
> I suspect the 3/5/6/7 cases are not really necessary. It seemed reasonable to just implement them, but I could _ASSERTE instead. My intent is to copy the argument stack size as reported by the ArgLocDesc

I would prefer an assert.

Reply:
I am looking at Vector3Interop it is hitting one of the new asserts.

I will rework this when I fix that failure.

Reply:
For the failing Vector3Interop case, the Vector3 argument is

```console
print *m_argLocDesc
(ArgLocDesc) $2 = {
  m_idxFloatReg = -1
  m_cFloatReg = 0
  m_idxGenReg = -1
  m_cGenReg = 0
  m_byteStackIndex = 28
  m_byteStackSize = 12
  m_hfaFieldSize = 0
```

I am surprised it is not 8-byte aligned...

Reply:
is it hfa<float>?

Reply:
Based on the alignment, it seems to be treated as HFA.  But I am surprised that `m_hfaFieldSize` is not `4` since it seems to be HFA.

I pushed a hack which inferred the alignment from the m_byteStackSize.  It feels wrong, but I don't really have a better idea.  I suspect it is sufficient to work for the current test coverage.

I wonder if `m_hfaFieldSize` should change to `m_stackAlignment`...

Reply:
so we expect hfaFieldSize to be set on stack passed arguments? It makes sense, maybe this line https://github.com/dotnet/runtime/blob/ad6a69735d61a1bc22da9aff15ebc5df94f63beb/src/coreclr/vm/callingconvention.h#L674-L683
prevents `setHFAFieldSize` from being set?

Reply:
I was thinking it needed to be set here 

https://github.com/dotnet/runtime/blob/ad6a69735d61a1bc22da9aff15ebc5df94f63beb/src/coreclr/vm/callingconvention.h#L715-L718

Reply:
yes, I think it will work.

Reply:
OK I revised using `m_hfaFieldSize`.  I think it is a correct fix.  I am resolving this .  PTAL again.